File: frontend/src/service/locationService.ts
import axios from "axios";

const API = "https://provinces.open-api.vn/api/v2";

const PROVINCES_KEY = "cached_provinces";
const WARDS_KEY = "cached_wards";

export const locationService = {

  async getProvinces() {
    try {
      const cached = localStorage.getItem(PROVINCES_KEY);

      if (cached) {
        return JSON.parse(cached);
      }

      const res = await axios.get(`${API}/p/`);

      localStorage.setItem(PROVINCES_KEY, JSON.stringify(res.data));

      return res.data;
    } catch (error) {
      console.error("Lỗi lấy tỉnh/thành:", error);
      return [];
    }
  },


  async getWardsByProvince(provinceCode: number) {
    try {
      const cacheKey = `${WARDS_KEY}_${provinceCode}`;

      const cached = localStorage.getItem(cacheKey);

      if (cached) {
        return JSON.parse(cached);
      }

      const res = await axios.get(`${API}/p/${provinceCode}?depth=2`);

      const wards = res.data?.wards || [];

      localStorage.setItem(cacheKey, JSON.stringify(wards));

      return wards;
    } catch (error) {
      console.error("Lỗi lấy phường/xã:", error);
      return [];
    }
  },


  clearCache() {
    localStorage.removeItem(PROVINCES_KEY);

    Object.keys(localStorage).forEach((key) => {
      if (key.startsWith(WARDS_KEY)) {
        localStorage.removeItem(key);
      }
    });
  },
};

File: mobile/lib/services/location_service.dart
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:shared_preferences/shared_preferences.dart';

class ProvinceResponse {
  final int code;
  final String name;

  ProvinceResponse({
    required this.code,
    required this.name,
  });

  factory ProvinceResponse.fromJson(Map<String, dynamic> json) {
    return ProvinceResponse(
      code: json['code'] ?? 0,
      name: json['name']?.toString() ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'code': code,
      'name': name,
    };
  }
}

class WardResponse {
  final int code;
  final String name;

  WardResponse({
    required this.code,
    required this.name,
  });

  factory WardResponse.fromJson(Map<String, dynamic> json) {
    return WardResponse(
      code: json['code'] ?? 0,
      name: json['name']?.toString() ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'code': code,
      'name': name,
    };
  }
}

class LocationService {
  static const String _baseUrl = 'https://provinces.open-api.vn/api/v2';

  static const String _provincesKey = 'cached_provinces';
  static const String _wardsKey = 'cached_wards';

  late final Dio _dio;

  LocationService() {
    _dio = Dio(
      BaseOptions(
        connectTimeout: const Duration(seconds: 10),
        receiveTimeout: const Duration(seconds: 10),
      ),
    );
  }

  Future<List<ProvinceResponse>> getProvinces({
    bool forceRefresh = false,
  }) async {
    final prefs = await SharedPreferences.getInstance();

    if (!forceRefresh) {
      final cached = prefs.getString(_provincesKey);

      if (cached != null && cached.isNotEmpty) {
        final List decoded = jsonDecode(cached);
        if (decoded.isNotEmpty) {
          return decoded
              .map((e) => ProvinceResponse.fromJson(e as Map<String, dynamic>))
              .toList();
        }
      }
    }

    final response = await _dio.get('$_baseUrl/p/');

    final List rawData = response.data ?? [];

    final provinces = rawData
        .map((e) => ProvinceResponse.fromJson(e as Map<String, dynamic>))
        .toList();

    await prefs.setString(
      _provincesKey,
      jsonEncode(provinces.map((e) => e.toJson()).toList()),
    );

    return provinces;
  }

  Future<List<WardResponse>> getWardsByProvince(
      int provinceCode, {
        bool forceRefresh = false,
      }) async {
    final prefs = await SharedPreferences.getInstance();
    final cacheKey = '${_wardsKey}_$provinceCode';

    if (!forceRefresh) {
      final cached = prefs.getString(cacheKey);

      if (cached != null && cached.isNotEmpty) {
        final List decoded = jsonDecode(cached);
        if (decoded.isNotEmpty) {
          return decoded
              .map((e) => WardResponse.fromJson(e as Map<String, dynamic>))
              .toList();
        }
      }
    }

    final response = await _dio.get(
      '$_baseUrl/p/$provinceCode',
      queryParameters: {
        'depth': 2,
      },
    );

    final List rawData = response.data?['wards'] ?? [];

    final wards = rawData
        .map((e) => WardResponse.fromJson(e as Map<String, dynamic>))
        .toList();

    await prefs.setString(
      cacheKey,
      jsonEncode(wards.map((e) => e.toJson()).toList()),
    );

    return wards;
  }

  Future<void> clearCache() async {
    final prefs = await SharedPreferences.getInstance();

    await prefs.remove(_provincesKey);

    final keys = prefs.getKeys();

    for (final key in keys) {
      if (key.startsWith(_wardsKey)) {
        await prefs.remove(key);
      }
    }
  }
}

final locationService = LocationService();